pkg/controller/elasticsearch/driver/upgrade_forced.go

```diff
@@ -0,0 +1,52 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// maybeForceUpgrade may attempt a forced upgrade of all podsToUpgrade if allowed to,
+// in order to unlock situations where the reconciliation may otherwise be stuck
+// (eg. no cluster formed, all nodes have a bad spec).
+func (d *defaultDriver) maybeForceUpgrade(actualPods []corev1.Pod, podsToUpgrade []corev1.Pod) (attempted bool, err error) {
+	attempted = false
+	if len(podsToUpgrade) == 0 {
+		return attempted, nil
+	}
+	if !shouldForceUpgrade(actualPods) {
+		return attempted, nil
+	}
+
+	attempted = true
+	log.Info("Performing a forced rolling upgrade since no Pod is ready",
+		"namespace", d.ES.Namespace, "es_name", d.ES.Name,
+		"pod_count", len(podsToUpgrade),
+	)
+	for _, pod := range podsToUpgrade {
+		if err := deletePod(d.Client, d.ES, pod, d.Expectations); err != nil {
+			return attempted, err
+		}
+	}
+	return attempted, nil
+}
+
+// shouldForceUpgrade returns true if all existing Pods can be safely upgraded,
+// without further safety checks.
+// In practice, we're OK to force-upgrade if all Pods are non-ready.
+// /!\ race condition: since the readiness is based on a cached value, we may allow
+// a forced rolling upgrade to go through based on out-of-date Pod data.
```

Comment:
An example of what may happen here:

1. User creates a cluster, and applies a new spec right after that
2. We rotate Pods while they are starting OR while they are already running and ready, but our cache does not reflect that yet
3. Pods are effectively restarted directly, in an "unsafe" way leading to downtime

I think this is generally fine since that cluster was just created anyway.
For already existing clusters, chances are we already see at least one Pod ready so we don't proceed with the forced upgrade.
If we see no Pods ready, there is still a chance that one becomes ready and we don't see it. Provided the cluster was already unavailable (because no Pods ready at a given time), I think it's OK to accidentally trigger another period of unavailability here.

If the operator gets disconnected from the apiserver for a long period, and happens to be in a situation where all pods are non-ready, it would also proceed with out-of-date readiness information. But only once, since the next upgrade would wait for expectations of the first one to be met.

I think it's OK to take that risk.